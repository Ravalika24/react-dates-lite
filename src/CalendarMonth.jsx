/* @flow */
import React from 'react';

import isSameDay from 'date-fns/is_same_day';
import startOfMonth from 'date-fns/start_of_month';
import endOfMonth from 'date-fns/end_of_month';
import eachDayOfInterval from 'date-fns/eachDayOfInterval';
import format from 'date-fns/format';
import subMonths from 'date-fns/sub_months';
import lastDayOfMonth from 'date-fns/last_day_of_month';
import getDate from 'date-fns/get_date';
import subDays from 'date-fns/sub_days';
import isSameMonth from 'date-fns/is_same_month';
import isBefore from 'date-fns/is_before';
import startOfDay from 'date-fns/start_of_day';
import getDay from 'date-fns/getDay';
import R from 'ramda';
import styled from 'styled-components';

import CalendarDay from './CalendarDay';

type Props = {|
  selectDate: Date => void,
  onHover: Date => void,
  month: Date,
  selectedDays: Date[],
  hoveredDates: Date[],
  allowedPastDates: boolean,
  colors: { [string]: string },
  classes: { [string]: string },
|};

const Week = styled.div`
  display: table-row;
`;

const Month = styled.div`
  width: 300px;
  display: table;
  border-collapse: collapse;
`;

const DayNameList = styled.div`
  display: flex;
  align-items: center;
  margin: 40px 0 10px;
`;

const DayName = styled.span`
  width: 43px;
  text-align: center;
`;

const MonthName = styled.span`
  height: 30px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const CalendarMonth = (props: Props) => {
  const {
    month,
    selectedDays,
    selectDate,
    onHover,
    hoveredDates,
    allowedPastDates,
    colors,
    classes,
  } = props;

  // TODO: standalone function
  const start = startOfMonth(month);
  const end = endOfMonth(month);
  const listOfDays = eachDayOfInterval({ start, end });
  const monthName = format(month, 'MMMM YYYY');
  const monthBefore = subMonths(month, 1);
  const lastDayPrevMonth = lastDayOfMonth(monthBefore);
  const toPrepend =
    getDay(start) === 0
      ? []
      : eachDayOfInterval({
          start: subDays(lastDayPrevMonth, getDay(start) - 1),
          end: lastDayPrevMonth,
        });

  const toRender = R.concat(toPrepend, listOfDays);
  const dayNames = R.take(7, toRender);

  return (
    <div>
      <MonthName>{monthName}</MonthName>

      <DayNameList>
        {R.map(
          day => <DayName key={day}>{format(day, 'dd')}</DayName>,
          dayNames,
        )}
      </DayNameList>

      <Month>
        {R.map(
          week => (
            <Week key={week}>
              {R.map(
                day => (
                  <CalendarDay
                    key={day}
                    isHidden={!isSameMonth(month, day)}
                    number={getDate(day)}
                    value={day}
                    selectDate={selectDate}
                    onHover={onHover}
                    isHovered={R.find(
                      selected => isSameDay(selected, day),
                      R.drop(1, hoveredDates),
                    )}
                    isSelected={R.find(
                      selected => isSameDay(selected, day),
                      selectedDays,
                    )}
                    isPast={
                      isBefore(day, startOfDay(new Date())) && !allowedPastDates
                    }
                    colors={colors}
                    classes={classes}
                  />
                ),
                week,
              )}
            </Week>
          ),
          R.splitEvery(7, toRender),
        )}
      </Month>
    </div>
  );
};

export default CalendarMonth;
