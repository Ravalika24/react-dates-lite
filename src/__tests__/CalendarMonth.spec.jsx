import React from 'react';
import { shallow } from 'enzyme';

// import getDate from 'date-fns/get_date';

import CalendarMonth from '../CalendarMonth';

describe('#CalendarMonth', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <CalendarMonth
        month={new Date(2018, 1, 1)}
        selectedDays={[]}
        disabledDays={[]}
        selectDate={jest.fn()}
        onHover={jest.fn()}
        hoveredDates={[]}
        allowedPastDates
        future
        colors={{}}
        classes={{}}
      />
    );
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});
