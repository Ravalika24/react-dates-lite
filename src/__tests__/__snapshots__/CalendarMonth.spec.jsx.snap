// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CalendarMonth should render correctly 1`] = `
<div
  className={undefined}
>
  <styled.span>
    February 2018
  </styled.span>
  <styled.div>
    <styled.span>
      Su
    </styled.span>
    <styled.span>
      Mo
    </styled.span>
    <styled.span>
      Tu
    </styled.span>
    <styled.span>
      We
    </styled.span>
    <styled.span>
      Th
    </styled.span>
    <styled.span>
      Fr
    </styled.span>
    <styled.span>
      Sa
    </styled.span>
  </styled.div>
  <styled.div>
    <styled.div>
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={true}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={28}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-01-27T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={true}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={29}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-01-28T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={true}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={30}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-01-29T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={true}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={31}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-01-30T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={1}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-01-31T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={2}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-01T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={3}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-02T23:00:00.000Z}
      />
    </styled.div>
    <styled.div>
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={4}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-03T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={5}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-04T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={6}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-05T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={7}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-06T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={8}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-07T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={9}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-08T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={10}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-09T23:00:00.000Z}
      />
    </styled.div>
    <styled.div>
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={11}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-10T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={12}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-11T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={13}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-12T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={14}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-13T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={15}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-14T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={16}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-15T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={17}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-16T23:00:00.000Z}
      />
    </styled.div>
    <styled.div>
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={18}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-17T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={19}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-18T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={20}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-19T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={21}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-20T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={22}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-21T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={23}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-22T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={24}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-23T23:00:00.000Z}
      />
    </styled.div>
    <styled.div>
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={25}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-24T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={26}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-25T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={27}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-26T23:00:00.000Z}
      />
      <CalendarDay
        classes={Object {}}
        colors={Object {}}
        isFuture={false}
        isHidden={false}
        isHovered={false}
        isPast={false}
        isSelected={false}
        number={28}
        onHover={[MockFunction]}
        selectDate={[MockFunction]}
        value={2018-02-27T23:00:00.000Z}
      />
    </styled.div>
  </styled.div>
</div>
`;
